packages/byods/src/data-source-service/index.ts

```diff
@@ -0,0 +1,160 @@
+/* eslint-disable no-useless-catch */
+/* eslint-disable no-console */
+/* eslint-disable class-methods-use-this */
+/* eslint-disable import/no-extraneous-dependencies */
+
+import {v4 as uuidv4} from 'uuid';
+import {decodeJwt, JWTPayload} from 'jose';
+import DataSourceClient from '../data-source-client/index';
+
+export default class DataSourceService {
+  private dataSourceClient: DataSourceClient;
+  private timer: any;
+
+  /**
+   * Creates an instance of DataSourceService.
+   * @param {DataSourceClient} dataSourceClient - The  dataSourceClient instance to use for DataSourceService.
+   * @example
+   * const dataSourceService = new DataSourceService(dataSourceClient);
+   */
+
+  constructor(dataSourceClient: DataSourceClient) {
+    this.dataSourceClient = dataSourceClient;
+  }
+
+  /**
+   * This method refreshes the DataSource token using dataSourceId and tokenLifetimeMinute
+   * @param {string} dataSourceId The id of data source
+   * @param {number} tokenLifetimeMinutes The Life time minutes for the data source. <=1440
+   * @returns {Promise<void>}
+   * @example
+   * const dataSourceService = new DataSourceService(dataSourceClient); Create an instance of dataSourceService by passing instance of dataSourceClient.
+   * dataSourceService.refreshDataSourceToken(12345, 60); Invoke refreshDataSourceToken method by invoking dataSourceId and tokenLifetimeMinutes.
+   */
+
+  public async refreshDataSourceToken(
+    dataSourceId: string,
+    tokenLifetimeMinutes?: number
+  ): Promise<void> {
+    try {
+      if (dataSourceId) {
+        if (tokenLifetimeMinutes) {
+          const reducedTokenLifetimeMinutes = await this.getReduceTokenLifetimeMinutes(
+            tokenLifetimeMinutes
+          );
+          this.startAutoRefresh(dataSourceId, reducedTokenLifetimeMinutes);
+        } else {
+          console.log('tokenLifetimeMinutes is not provided, setting up default value');
+          const reducedTokenLifetimeMinutes = await this.getReduceTokenLifetimeMinutes(60);
+          this.startAutoRefresh(dataSourceId, reducedTokenLifetimeMinutes);
+        }
+      } else {
+        console.log('Required field is missing, Please pass dataSourceId.');
+      }
+    } catch (error) {
+      console.log('Encountered some error, error is', error);
+    }
+  }
+
+  /**
+   * This Private method will be used to reduce the tokenLifetimeMinutes from 5% to 10%
+   * @param {number} tokenLifetimeMinutes The Life time minutes for the data source. <=1440
+   * @returns {number} It returns a reduced token lifetime minutes.
+   * @example
+   */
+
+  private async getReduceTokenLifetimeMinutes(tokenLifetimeMinutes: number): Promise<number> {
+    // Below code will generate a random percentage between 5% and 10%
+    const randomPercentage = Math.random() * (10 - 5) + 7;
```

Comment:
Check the logic. not sure this gives always give 5-10

Reply:
**_Updated the logic!_**